src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs

```diff
@@ -48,27 +25,30 @@ internal class KafkaSourceStageLogic<K, V> : GraphStageLogic
         private readonly TaskCompletionSource<NotUsed> _completion;
         private readonly CancellationTokenSource _cancellationTokenSource;
 
-        public KafkaSourceStageLogic(KafkaSourceStage<K, V> stage, Attributes attributes, TaskCompletionSource<NotUsed> completion) : base(stage.Shape)
+        public SingleSourceStageLogic(SourceShape<TMessage> shape, ConsumerSettings<K, V> settings, 
+            ISubscription subscription, Attributes attributes, 
+            TaskCompletionSource<NotUsed> completion, IMessageBuilder<K, V, TMessage> messageBuilder) 
+            : base(shape)
         {
-            _settings = stage.Settings;
-            _subscription = stage.Subscription;
+            _settings = settings;
+            _subscription = subscription;
             _completion = completion;
             _cancellationTokenSource = new CancellationTokenSource();
 
             var supervisionStrategy = attributes.GetAttribute<ActorAttributes.SupervisionStrategy>(null);
             _decider = supervisionStrategy != null ? supervisionStrategy.Decider : Deciders.ResumingDecider;
 
-            SetHandler(stage.Out, onPull: () =>
+            SetHandler(shape.Outlet, onPull: () =>
             {
                 try
                 {
                     var message = _consumer.Consume(_cancellationTokenSource.Token);
                     if (message == null) // No message received, or consume error occured
                         return;
 
-                    if (IsAvailable(stage.Out))
+                    if (IsAvailable(shape.Outlet))
                     {
-                        Push(stage.Out, message);
+                        Push(shape.Outlet, messageBuilder.CreateMessage(message, _consumer));
```

Comment:
Using generic `IMessageBuilder` to build downstream message.
The tricky part here is passing `_consumer` instance to builder. Alpakka does not do that. Instead, they use GraphStage's `materializer` property to get consumer actor reference and use it in message builder like this:

```scala
  override protected def logic(shape: SourceShape[CommittableMessage[K, V]]): GraphStageLogic with Control =
    new SingleSourceLogic[K, V, CommittableMessage[K, V]](shape, settings, subscription)
    with CommittableMessageBuilder[K, V] {
      override def metadataFromRecord(record: ConsumerRecord[K, V]): String = _metadataFromRecord(record)
      override def groupId: String = settings.properties(ConsumerConfig.GROUP_ID_CONFIG)
      lazy val committer: KafkaAsyncConsumerCommitterRef = {
        val ec = materializer.executionContext
        new KafkaAsyncConsumerCommitterRef(consumerActor, settings.commitTimeout)(ec)
      }
    }
```

We do not use any consumer actor right now, neither Akka.Streams gives `materializer` property - or maybe I did not find it - so to allow builder to execute commit action, I am passing consumer instance directly.

Reply:
@IgorFedchenko the materializer might be passed in as a Scala `implicit` - which makes the code somewhat hard to read. I'll take a look and see how they're doing this.

Reply:
Or maybe you might have some other ideas